### DISCUSSION TITLE
Principles for Outfit Organization with Series

### DISCUSSION THREAD

**Discussion by Daeridanii1** [endless-sky/endless-sky#10335] [Balance, Mechanics, and Interactions]
## Introduction
series.txt allows us to define "outfit series", which are ordered lists of outfits that are displayed _in that order_ in the outfitter. So far, this is used only by Human engines (as of b8a5631) to display them in an organized fashion, but this system can be applied to any outfits to make them well-organized in their categories.

After discussions on Discord, I wrote up this post (collating the discussions therin) to have a more permanent space to discuss how, if at all, we might want to more broadly apply series outfit organization to ES's content. The following principles are mostly my own suggestions, provided as a starting point for discussion: ideally, there should be an established consensus for what constitutes standard outfit series before any large-scale effort at serializing existing content is undertaken.

## Why?

If applied at scale, this could help make the player experience of navigating different outfitters far more consistent. For example, Human batteries are more or less listed after reactors but before fuel cells, while Coalition batteries, reactors, and fuel cells are entirely interleaved. Ideally, these should be consistent: if reactors are listed first at one outfitter, they should be listed the same relative to other outfits at another outfitter. This makes for a better user experience.

## Principles for Series Organization of Outfits
By order of priority
1. "Inversions" of specific outfits should be adjacent. e.g. Outfit Expansions and Cargo Expansions
2. "Dependents" that require an outfit of a specific type should be listed immediately after all outfits of that specific type. e.g. Bunrodea Shield Relay Boosters and Limiters should be listed immediately after all Bunrodea Shield Relays.
3. Within an series of outfits primarily distinguished by size, the series should be ordered smallest to largest. 

## Established Precedents
1. Engines: combined --> steering --> thruster --> reverse

## Proposed Precedents, per category

(Items in parentheses) refer to general ordering within a specific series
**Items in bold** are universal; only one per category

_Guns_: None specifically
_Turrets_: None specifically
_Secondary Weapons_: Fixed --> Turreted
_Ammunition_: Ammo --> Ammo Storage
_Systems_: **Drives** --> Regen (Shields, Hull Repair) --> Cooling --> Fuel (Tanks, Ramscoops) --> _Info_ (Scanners/Jamming) --> **Passenger** --> **Utility** (everything else, Cargo/Outfit Expansions)
_Power_: Generators (Low-Power, Reactors, Solar) --> Batteries
_Engines_: See established precedents
_Hand to Hand_: Offensive --> Defensive --> Fortifications
_Unique_: Useful before non-useful
_Special_: Does not need serialization
_Licenses_: Pilots --> Others (ascending in "power")

### Establishing series numbers
In each category, if there is a universal high-priority series, it has number 0
If there is a universal low-priority series, it has number 9999

For others, 4 digits:
Each precedent division (outfit type) within a category defines the first two digits within that category
The remaining two digits are defined by the species that predominantly uses that outfit
00 Humans
01 Hai
02 Remnant
03 Korath
04 Ka'het
05 Gegno
06 Coalition
07 Wanderers
08 Bunrodea
09 Pug
10 Quarg
11 _Incipias_
12 _Successors_
13 _Avgi_
99 Undefined/Other

New species should simply be appended to this list as necessary by order of addition.

### Example:
A programmatically-generated `series.txt` and the Python script that generated it. Sorry if my code is ugly, but hey it works.
https://github.com/Daeridanii1/outfit-serial-list-generator/blob/master/series.txt

Edits:
<details>
 * 3 --> 4 digit series numbers, mOctave suggestions
 * Removed references to "serialization" for disambiguation
</details>

**Comment by mOctave** [comment]
Looks good! I'm glad somebody's finally thinking about serializing more than just a few engines. A few thoughts:

First, I'd probably sort hand-to-hand according to purpose (offensive, defensive, fortification), although not necessarily by amount of actual damage dealt.

For licenses, I think ideally the pilot's license should come first, then sorted by species. Beyond that, you're right that I don't think it matters, although we could sort by organization (eg Republic/FW/Deep), and then by license "power" (capital ship license after general purpose license).

I'm also not quite sure how I feel about lumping cargo/outfit expansions in with bunk rooms. I feel like the utility section could be broken out a bit more.

Also, would Drak/Archon stuff be 11, or 9999? I'd probably go with 11, since they have some overlap with Quarg.

**Reply by Daeridanii1** [reply]
I agree completely with your thoughts on h2h and licenses, I think those are very reasonable.

The _Utility_ series will no doubt always end up being eclectic; "Systems" contains so much unique equipment that's hard to strictly categorize with anything else. My reasoning for putting expansions and bunk rooms together is that they both represent numerical changes to a ship's functional role. However, I could definitely see bunk rooms in a series with luxury accommodations and brigs as part of a "passenger" series. 

As for Archon equipment, I think it would be best categorized under 9999, because it is meant to be permanently legitimately inaccessible. Thus, it will only every be visible with save editing or plugins. In light of this, I think it's appropriate to put it at the end with other normally inaccessible outfits so that accessible content is always prioritized.

**Comment by Zireael07** [comment]
small nitpick: in the development context serialization usually means saving to disk, so I was super confused by the title and contents

**Comment by Tangle10** [comment]
As it stands, for the record, applying this to the few existing series would take the Human series from Human: Ion/Plasma/Atomic/Afterburner 110/120/130/140 to... 0000/0100/0200/0300, I think? I'm not good at parsing zero-indexing.
